abci/src/codec.rs

```diff
@@ -0,0 +1,166 @@
+//! Encoding/decoding mechanisms for ABCI requests and responses.
+//!
+//! Implements the [Tendermint Socket Protocol][tsp].
+//!
+//! [tsp]: https://docs.tendermint.com/master/spec/abci/client-server.html#tsp
+
+use crate::Result;
+use bytes::{Buf, BufMut, BytesMut};
+use prost::Message;
+use std::io::{Read, Write};
+use std::marker::PhantomData;
+use tendermint_proto::abci::{Request, Response};
+
+/// The maximum number of bytes we expect in a varint. We use this to check if
+/// we're encountering a decoding error for a varint.
+pub const MAX_VARINT_LENGTH: usize = 16;
+
+/// The server receives incoming requests, and sends outgoing responses.
+pub type ServerCodec<S> = Codec<S, Request, Response>;
+
+#[cfg(feature = "client")]
+/// The client sends outgoing requests, and receives incoming responses.
+pub type ClientCodec<S> = Codec<S, Response, Request>;
+
+/// Allows for iteration over `S` to produce instances of `I`, as well as
+/// sending instances of `O`.
+///
+/// `S` must implement [`std::io::Read`] and [`std::io::Write`].
```

Comment:
Instead of documenting it, `S` could be constrained here.

Reply:
The pattern of constraining struct type params when not absolutely necessary* is discouraged in the Rust ecosystem (most notably in the standard library): https://github.com/rust-lang/api-guidelines/issues/6

As such, the current implementation looks just fine to me.

\* A struct should only have trait bounds on its type params if it needs to refer to a type member defined in a trait, eg.

```rust
trait Foo {
  type Bar;
}

struct Toto<F: Foo> {
  tata: F::Bar
}
```

Reply:
Thanks for the pointers in the guidelines. It is sensible to not constraint the params here, but I'd drop the doc hint for it then.

Reply:
Cool, I've removed the comment :+1: